zeebe/engine/src/test/java/io/camunda/zeebe/engine/processing/bpmn/activity/listeners/task/TaskListenerTest.java

```diff
@@ -944,6 +945,162 @@ public void shouldAssignAndCompleteTaskAfterTaskListenerRejectsTheCompletion() {
+        UserTaskIntent.ASSIGNED,
+        userTask -> Assertions.assertThat(userTask).hasAssignee("me").hasAction("assign"));
+  }
+
+  @Test
+  public void shouldRetryUserTaskClaimCommandAfterExtractValueErrorIncidentResolution() {
+    testUserTaskCommandRetryAfterExtractValueError(
+        ZeebeTaskListenerEventType.assigning,
+        "claim_listener_var_name",
+        "expression_claim_listener_2",
+        userTask -> userTask.withAssignee("me").claim(),
+        UserTaskIntent.ASSIGNED,
+        userTask -> Assertions.assertThat(userTask).hasAssignee("me").hasAction("claim"));
+  }
+
+  private void testUserTaskCommandRetryAfterExtractValueError(
+      final ZeebeTaskListenerEventType eventType,
+      final String variableName,
+      final String variableValue,
+      final Consumer<UserTaskClient> userTaskAction,
+      final UserTaskIntent expectedIntent,
+      final Consumer<UserTaskRecordValue> assertion) {
+
+    // given
+    final long processInstanceKey =
+        createProcessInstance(
+            createUserTaskWithTaskListeners(
+                eventType, LISTENER_TYPE, "=" + variableName, LISTENER_TYPE + "_3"));
+
+    // when: perform the user task action
+    userTaskAction.accept(ENGINE.userTask().ofInstance(processInstanceKey));
+
+    // complete the first task listener job
+    ENGINE.job().ofInstance(processInstanceKey).withType(LISTENER_TYPE).complete();
+
+    // then: expect an incident due to missing variable
+    final var incidentRecord =
+        RecordingExporter.incidentRecords(IncidentIntent.CREATED)
+            .withProcessInstanceKey(processInstanceKey)
+            .getFirst();
+
+    Assertions.assertThat(incidentRecord.getValue())
+        .hasProcessInstanceKey(processInstanceKey)
+        .hasErrorType(ErrorType.EXTRACT_VALUE_ERROR)
+        .hasErrorMessage(
+            """
+                Expected result of the expression '%s' to be 'STRING', but was 'NULL'. \
+                The evaluation reported the following warnings:
+                [NO_VARIABLE_FOUND] No variable found with name '%s'"""
+                .formatted(variableName, variableName));
+
+    // when: fix the missing variable and resolve the incident
+    ENGINE
+        .variables()
+        .ofScope(processInstanceKey)
+        .withDocument(Map.of(variableName, variableValue))
+        .update();
+    ENGINE.incident().ofInstance(processInstanceKey).withKey(incidentRecord.getKey()).resolve();
+
+    // complete the retried task listener job and remaining task listeners
+    completeRecreatedJobWithType(ENGINE, processInstanceKey, LISTENER_TYPE);
+    completeJobs(processInstanceKey, variableValue, LISTENER_TYPE + "_3");
+
+    // then
+    assertTaskListenerJobsCompletionSequence(
+        processInstanceKey,
+        mapToJobListenerEventType(eventType),
+        LISTENER_TYPE,
+        LISTENER_TYPE, // re-created task listener job
+        variableValue,
+        LISTENER_TYPE + "_3");
```

Comment:
💭 I was a bit confused for a second with the listeners names. I understand what the test is doing 👍 it is not straight forward to follow. I am not sure how to improve it though - if there is a better way 🤔